### PR TITLE
KAFKA-2479: Transient exception

### PR DESCRIPTION
**Pull request by Ishiihara**
Transient exception handling.

### CODE REVIEW

**Comment by ewencp** [comment]
@Ishiihara KAFKA-2482 has been merged, so you could cherry-pick/merge this to get a clean PR including only your changes.

File: copycat/api/src/main/java/org/apache/kafka/copycat/errors/TransientException.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ **/
+
+package org.apache.kafka.copycat.errors;
+
+public class TransientException extends CopycatException {
+
+  public TransientException(String s) {
```
Comment (ewencp) [inline copycat/api/src/main/java/org/apache/kafka/copycat/errors/TransientException.java:20]: checkstyle is complaining about wrong indent levels

File: copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java [hunk 2]
```diff
@@ -59,6 +60,22 @@ public void offset(Map<TopicPartition, Long> offsets) {
     }
 
     /**
+     * Get the topic partition that throws Transient exception during message delivery
+     * @return the topic partition that has error
+     */
+    public TopicPartition erorrTopicPartition() {
```
Comment (ewencp) [inline copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java:66]: Typo: erorr -> error

File: copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java [hunk 3]
```diff
@@ -59,6 +60,22 @@ public void offset(Map<TopicPartition, Long> offsets) {
...
+        return errorTopicPart;
```
Comment (ewencp) [inline copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java:67]: See note I left in the KAFKA-2480 PR. I didn't notice this with the commit that added the `offsets`, but I think we want to hide the accessors

File: copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java [hunk 4]
```diff
@@ -59,6 +60,22 @@ public void offset(Map<TopicPartition, Long> offsets) {
...
+    }
+
+    /**
+     * Set the error topic partition
+     * @param errorTopicPart the topic partition with error
+     */
+    public void errorTopicPartition(TopicPartition errorTopicPart) {
```
Comment (ewencp) [inline copycat/api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java:74]: I assume this is the method that is supposed to be used by connectors. Can the comment explain in more detail how it is supposed to work? What if you have multiple TopicPartitions with errors?

File: copycat/runtime/src/test/java/org/apache/kafka/copycat/runtime/WorkerSinkTaskTest.java [hunk 5]
```diff
@@ -349,6 +358,52 @@ public Object answer() throws Throwable {
...
+        Properties taskProps = new Properties();
+        expectInitializeTask(taskProps);
+
+        sinkTask.put(EasyMock.anyObject(Collection.class));
+        EasyMock.expectLastCall().andThrow(new TransientException("Transient exception from SinkTask"));
+
+        EasyMock.expect(consumer.assignment()).andReturn(
+            new HashSet<>(Arrays.asList(TOPIC_PARTITION, TOPIC_PARTITION2)));
+
+        consumer.pause(TOPIC_PARTITION);
```
Comment (ewencp) [inline copycat/runtime/src/test/java/org/apache/kafka/copycat/runtime/WorkerSinkTaskTest.java:372]: This test doesn't seem to correspond to the code -- the context object provides one TopicPartition that you can set as the one causing the error, but then this pauses two partitions.

**Comment by guozhangwang** [comment]
@Ishiihara Liquan could you rebase if still relevant?

**Comment by b-goyal** [comment]
Transient exception handling.

**Comment by b-goyal** [comment]
Transient exception handling.

**Comment by b-goyal** [comment]
Transient exception handling.

**Pull request closed without merging**